### PR TITLE
Show missing minions in lists & nodegroups

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Marks non-existent nodes specified in a list or as part of a node group as failed when called via the `salt` command.

### What issues does this PR fix or reference?
#41326

### Previous Behavior
Non-existent nodes were silently ignored.

### New Behavior
Commands are not sent to non-existent nodes but they are included in the final output and marked as failed.

### Tests written?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Comment by rallytime** [comment]
@garethgreenaway Looks like there are some related test failures here: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/13756/#showFailuresLink

File: salt/master.py [hunk 1]
```diff
@@ -1853,11 +1853,13 @@ def publish(self, clear_load):
 
         # Retrieve the minions list
         delimiter = clear_load.get(u'kwargs', {}).get(u'delimiter', DEFAULT_TARGET_DELIM)
-        minions = self.ckminions.check_minions(
+        _res = self.ckminions.check_minions(
             clear_load[u'tgt'],
             clear_load.get(u'tgt_type', u'glob'),
             delimiter
         )
+        minions = _res.get('minions', list())
```
Comment (thatch45) [inline salt/master.py]: I thought `[]` was always faster than `list()`? Am I mistaken?

File: salt/utils/minions.py [hunk 2]
```diff
@@ -201,7 +201,8 @@ def _check_glob_minions(self, expr, greedy):  # pylint: disable=unused-argument
         '''
         Return the minions found by looking via globs
         '''
-        return fnmatch.filter(self._pki_minions(), expr)
+        return {'minions': fnmatch.filter(self._pki_minions(), expr),
+                'missing': list()}
```
Comment (thatch45) [inline salt/utils/minions.py]:
```
In [4]: timeit.timeit('[]', number=100000)
Out[4]: 0.003401219997613225

In [5]: timeit.timeit('list()', number=100000)
Out[5]: 0.016737269001168897

In [6]: timeit.timeit('dict()', number=100000)
Out[6]: 0.014586324999982025

In [7]: timeit.timeit('{}', number=100000)
Out[7]: 0.004838772998482455
```
Basically using `list()` and `dict()` are slow and should be avoided for simple use cases

**Comment by thatch45** [comment]
I am on the fence on this one. I think it adds some great insight, and makes the interface more flexible. But I am wary about the fact that it changes an interface like that. Granted I don't think this gets used a ton outside of Salt, and it is ok to change an interface like this in a major release, and it is an easy fix for anyone using CkMinions from the outside.

In the end I am more in favor than against, I like having the insight.

Oh! Does this change what the load.p file looks like for the returner? That could cause some chaos....

File: salt/utils/minions.py (hunk 2 above)
Comment (garethgreenaway) [reply salt/utils/minions.py]: :+1: I have replaced all the occurrences of `list()` with `[]`

Comment (thatch45) [reply salt/utils/minions.py]: thanks!

**Comment by garethgreenaway** [comment]
@thatch45 Running a simple test using the Slack returner, the data is that the returner receives only includes the minions that exist.

**Comment by rallytime** [comment]
re-run lint

**Comment by thatch45** [comment]
Sounds good, I am good to go with this since we know it is not going to break the load.p file

**Review by thatch45 — APPROVED** [review]

**Pull request merged**